### PR TITLE
Refactor and implement JWT auth

### LINKED ISSUE #19: Refactor the codebase

### LINKED ISSUE #49: Use JWT authentication
**__Task Description__**
TBA

**__Technical Considerations__**
TBA

**__Additional Considerations__**
TBA

### PR DESCRIPTION
**Pull request by alvitoraidhy**
Closes #19 and #49. Going to need everyone's approval for this because a lot was changed and/or refactored.

### CODE REVIEW

File: buddies-server/app.js [hunk 1]
```diff
@@ -1,23 +1,22 @@
-var createError = require("http-errors");
```
Comment (elliotwutingfeng) [inline buddies-server/app.js:1]: Good. We should stop using `var` as far as possible.

File: buddies-server/lib/utils.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// ref: https://stackoverflow.com/a/48733909
+function compose(middleware) {
+  if (!middleware.length) {
+    return function(_req, _res, next) { next(); };
+  }
+
+  var head = middleware[0];
+  var tail = middleware.slice(1);
```
Comment (elliotwutingfeng) [inline buddies-server/lib/utils.js]: This could be `const`

File: buddies-server/routes/users.js [hunk 3]
```diff
@@ -1,99 +1,94 @@
...
-        res.status(200).json("user has been followed");
+        return res.status(200).json("user has been followed");
       } else {
-        res.status(403).json("you allready follow this user");
+        return res.status(403).json("you already follow this user");
       }
-    } catch (err) {
-      res.status(500).json(err);
+    } else {
+      returnres.status(403).json("you cant follow yourself");
```
Comment (elliotwutingfeng) [inline buddies-server/routes/users.js]: ⛔ There is a typo here! `returnres` should be `return res`

**Review by elliotwutingfeng — CHANGES_REQUESTED** [review]
Typo found in `users.js` that could lead to runtime error. Apart from that the application runs well on my end.

File: buddies-server/lib/utils.js (hunk 2 above)
Comment (alvitoraidhy) [reply buddies-server/lib/utils.js]: I agree. I took the snippet without changing anything.

Comment (alvitoraidhy) [reply buddies-server/lib/utils.js]: Pushed a new commit to resolve this

File: buddies-server/routes/users.js (hunk 3 above)
Comment (alvitoraidhy) [reply buddies-server/routes/users.js]: Oops. My bad

Comment (alvitoraidhy) [reply buddies-server/routes/users.js]: Pushed a new commit to resolve this

**Comment by alvitoraidhy** [comment]
> Typo found in `users.js` that could lead to runtime error. Apart from that the application runs well on my end.

Thanks. The issue should be fixed.

File: buddies-client/src/components/EventPopup.js [hunk 4]
```diff
@@ -117,9 +145,9 @@ export default function EventPopup({ currentEventId, togglePopup }) {
                   {eventData.guests.length}
                 </div>
               </i>
-              {isLoggedIn ? (
+              {user ? (
                 <div>
-                  {eventData.guests.includes(user._id) ? (
+                  {eventData.guests.includes(user.id) ? (
```
Comment (elliotwutingfeng) [inline buddies-client/src/components/EventPopup.js]: `_id` is now being used in the backend but the frontend is still using `id`.

File: buddies-server/routes/auth.js [hunk 5]
```diff
@@ -33,37 +32,78 @@ router.post(
 
       // save user
       const user = await newUser.save();
-      res.status(200).json(user);
+
+      const token = generateToken({
+        id: user._id,
```
Comment (elliotwutingfeng) [inline buddies-server/routes/auth.js]: `_id` is now being used in the backend but the frontend is still using `id`.

**Review by elliotwutingfeng — CHANGES_REQUESTED** [review]
Looks good now, except for the use of `_id` in the backend vs `id` in the frontend. It can be resolved by renaming all the `id`s in frontend to `_id`.

File: buddies-server/routes/auth.js (hunk 5 above)
Comment (alvitoraidhy) [reply buddies-server/routes/auth.js]: > `_id` is now being used in the backend but the frontend is still using `id`.

Yes. Actually I didn't really like _id as an attribute name although I knew it was MongoDB's default for id, that's why I changed it to id on the frontend. But I guess consistency is more important. What do you think?

Comment (alvitoraidhy) [reply buddies-server/routes/auth.js]: Nevermind. Pushed a commit to resolve this

**Comment by alvitoraidhy** [comment]
> Looks good now, except for the use of `_id` in the backend vs `id` in the frontend. It can be resolved by renaming all the `id`s in frontend to `_id`.

Done. Thanks

**Review by elliotwutingfeng — APPROVED** [review]

**Comment by alvitoraidhy** [comment]
Merged to "chat-conversation" branch

**Pull request closed without merging**